molecules/main.py

```diff
@@ -0,0 +1,293 @@
+                      default=None,
+                      help='The Google Cloud Project ID when running with '
+                           'the `--cloud` option')
+  parser.add_argument('--region',
+                      type=str,
+                      default='us-central1',
+                      help='The Google Compute Engine region for the Dataflow '
+                           'and Cloud ML Engine jobs')
+
+  args, pipeline_args = parser.parse_known_args()
+
+  # 1) Input validation
+  if args.total_data_files < 1 and args.total_data_files != -1:
+    print('Error: --total-data-files must be >= 1 or -1 to use all')
+    sys.exit(1)
+
+  if args.total_data_files == -1:
+    args.total_data_files = None
+
+  if args.eval_percent <= 0 or args.eval_percent >= 100:
+    print('Error: --eval-percent must be in the range (0-100)')
+    sys.exit(1)
+
+  beam_options = None
+  if args.cloud:
+    if args.project is None:
+      print('error: argument --project is required with --cloud')
+      sys.exit(1)
+
+    if not args.work_dir.startswith('gs://'):
+      print('error: the --work-dir location must be in Google Cloud Storage '
+            'when running with --cloud')
+      sys.exit(1)
+
+    setup_file = os.path.join(os.path.dirname(__file__), 'setup.py')
+    beam_args = {
+        'job_name': 'cloudml-sample-molecules-preprocess-' + timestamp(),
+        'runner': 'DataflowRunner',
+        'project': args.project,
+        'temp_location': os.path.join(args.work_dir, 'beam_temp'),
+        'save_main_session': True,
+        'setup_file': os.path.abspath(setup_file),
+        'region': args.region,
+    }
+    beam_options = pipeline_options.PipelineOptions(flags=[], **beam_args)
+
+  print()
+  print('Working directory: ' + args.work_dir)
+  print()
+
+  # 2) Data extraction
+  print('Data extraction')
+  raw_data_dir = os.path.join(args.work_dir, 'raw_data')
+  raw_data_files_pattern = data_extractor.extract(
+      args.total_data_files, raw_data_dir)
+  print('Data extraction done')
+  print()
+
+  # 3) Preprocessing
+  print('Preprocessing')
+  train_input_fn, eval_input_fn, serving_input_fn = preprocess.run(
+      INPUT_SCHEMA,
+      LABELS,
+      PubChemToInputData(raw_data_files_pattern),
+      normalize_inputs,
+      beam_options=beam_options,
+      eval_percent=args.eval_percent,
+      temp_dir=args.work_dir)
+  print('Preprocessing done')
+  print()
+
+  # 4) Training and evaluating
+  print('Training and evaluating')
+
+  model_dir = os.path.join(args.work_dir, 'model')
+  run_config = tf.estimator.RunConfig()
+  run_config = run_config.replace(model_dir=model_dir)
+  estimator = make_estimator(run_config)
+
+  if not args.cloud:
+    # Running locally
+    trainer.train_and_evaluate(
+        train_input_fn,
+        eval_input_fn,
+        serving_input_fn,
+        estimator,
+        train_max_steps=args.train_max_steps)
+
+  else:
+    # Running in Cloud ML Engine. Pickle everything to pass in to the trainer
```

Comment:
there might be a simpler way to do this than having to pickle the functions?  the `gcloud mlengine jobs submit ...` command sends the whole trainer package according to your setup.py to a staging area and then that package gets deployed to CMLE.  So if these functions are already included in the package they should be available to the CMLE training job.